### DISCUSSION TITLE
Custom SSE endpoints?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3538] [Q&A]
I have a few use cases where it would be helpful to create custom endpoints that return an initial dump of data and then remain open as an SSE connection to stream updates. Can you tell me your thoughts on this and how you might approach it? (golang)

**Comment by ganigeorgiev** [comment]
The question is too vague and I'm not really sure what to recommend as it will depend on your specific use case.

An SSE endpoint is not much different than a regular route. The main differences are that:
- You need to make the route "blocking" (eg. using a `for {}` loop).
- Messages are written to the client using the response `Write()` and `Flush()` methods.

For an example you can check [apis/realtime.go](https://github.com/pocketbase/pocketbase/blob/94253f0dd5431e6a5676432537182bd2b0f2adae/apis/realtime.go).

But again it will depend on your actual use case - what is the data you want to return and who should have access to it.

---

Another possible approach could be reuse the existing subscriptions handling and send messages to a custom realtime topic.

On the server-side, you can access the subscribers using [`app.SubscriptionsBroker().Clients()`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.19.0/tools/subscriptions#Broker).
Each `Client` has a convenient [`Send(m subsctions.Message)`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.19.0/tools/subscriptions#Client) method.

On the client-side, you can use `pb.realtime.subscribe("YOUR_TOPIC", callback)`.

**Reply by benallfree** [reply]
As PocketHost implements regional support, `instances` are going to have an `edge` hostname identifier to say which machine around the globe that particular instance lives on. So, the edge servers need to monitor the central `instances` collection for changes to records matching the `edge` hostname of that edge server. It would be nice to create a realtime endpoint like `/api/edge/:edgeId` that could begin with an initial dump for that edge ID and then send subsequent updates as qualifying instances changed `active` status or whatever else I wanted to monitor.

**Reply by ganigeorgiev** [reply]
Unfortunately I'm still not eniterly sure that I understand the use case, but if your data that you want to track is already stored as collection records (aka. is not some ephemeral/temp messages), then I'd recommend using the existing collection subscription handling (_`filter`, `expand` and other request parameters support for the realtime subscriptions are planned for the next release_).

If the _"initial dump for that edge ID"_ is the only reason you want a dedicated SSE endpoint, then in my opinion is not worth the effort and it would be a lot easier to just send a regular `getList` request before the subscription to retrieve the data you want.

**Reply by benallfree** [reply]
I think you understand correctly. Edge servers will have Caddy running a firewall that will reject host headers that don't match its list of known PocketBase instances running on that server, or instances that have been disabled for whatever reason. So Caddy needs to get realtime updates from the mothership as instance records belonging to that edge server are updated/added/removed.

> If the "initial dump for that edge ID" is the only reason you want a dedicated SSE endpoint, then in my opinion is not worth the effort and it would be a lot easier to just send a regular getList request before the subscription to retrieve the data you want.

Yeah I've always had a theoretical problem with that. Correct me if I'm wrong, but couldn't I miss record updates between the `getList` call and the realtime subscription? It makes more sense to me to subscribe first, then call `getList` and possibly risk a few duplicates. I wonder if you'd be willing to consider a realtime `initial dump` option once you implement request params for realtime. That would be very helpful imho.

**Reply by ganigeorgiev** [reply]
>  It makes more sense to me to subscribe first, then call getList and possibly risk a few duplicates.

Yes, that sounds reasonable.

> I wonder if you'd be willing to consider a realtime initial dump option once you implement request params for realtime. 

I'm not sure how this will work as its behavior would be ambiguous, eg. what would be return result if you have 1M records?
While I agree that it would be more convenient, I think it will complicate the already slightly-messy realtime subscription APIs. Firing a subscription event only on change it is a lot more easy to reason and for the cases where the initial data is needed the alternative in my opinion with the dedicated list call is not that bad as users this way have more control over how many items, sort, etc. they want to fetch.

**Reply by benallfree** [reply]
> I'm not sure how this will work as its behavior would be ambiguous, eg. what would be return result if you have 1M records?

I think 1M records is rare enough that it could default to an initial dump and the user could opt out if needed. Or maybe PocketBase has some kind of safety limit.

I would like realtime to work like a "live query" where PocketBase tells me if any records are added or updated. That's how I'm using it anyway, in every single case.

Having a gap exist where records could have possibly been updated between the initial query and the live stream is theoretically uncomfortable even if it practically never happens.
